File: cli/cmd/llm.go
package cmd

import (
	"fmt"
	"time"

	"github.com/spf13/cobra"

	"openpa.local/cli/internal/output"
)

func newLLMCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "llm",
		Short: "Manage LLM providers, models, and model groups",
	}
	cmd.AddCommand(
		newLLMProvidersCmd(),
		newLLMModelGroupsCmd(),
		newLLMDeviceCodeCmd(),
	)
	return cmd
}

func newLLMProvidersCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "providers",
		Short: "List, configure, and inspect LLM providers",
	}
	cmd.AddCommand(
		&cobra.Command{
			Use:   "list",
			Short: "List configured and available providers",
			RunE: func(c *cobra.Command, _ []string) error {
				if err := requireToken(); err != nil {
					return err
				}
				providers, err := state.Client.ListLLMProviders(c.Context())
				if err != nil {
					return err
				}
				if state.Output.JSON {
					return output.PrintJSON(providers)
				}
				t := output.NewTable(state.Output, "NAME", "DISPLAY", "CONFIGURED", "MODELS", "ACTIVE_AUTH")
				for _, p := range providers {
					t.AddRow(
						stringField(p, "name"),
						stringField(p, "display_name"),
						boolField(p, "configured", false),
						numField(p, "model_count"),
						stringField(p, "active_auth_method"),
					)
				}
				t.Render()
				return nil
			},
		},
		&cobra.Command{
			Use:   "models <provider>",
			Short: "List models for a provider",
			Args:  cobra.ExactArgs(1),
			RunE: func(c *cobra.Command, args []string) error {
				if err := requireToken(); err != nil {
					return err
				}
				out, err := state.Client.GetProviderModels(c.Context(), args[0])
				if err != nil {
					return err
				}
				if state.Output.JSON {
					return output.PrintJSON(out)
				}
				models, _ := out["models"].([]any)
				t := output.NewTable(state.Output, "ID", "NAME")
				for _, m := range models {
					mm, ok := m.(map[string]any)
					if !ok {
						continue
					}
					t.AddRow(stringField(mm, "id"), stringField(mm, "name"))
				}
				t.Render()
				return nil
			},
		},
		newLLMConfigureCmd(),
		newLLMDeleteConfigCmd(),
	)
	return cmd
}

func newLLMConfigureCmd() *cobra.Command {
	var apiKey, authMethod string
	var extraJSON string
	cmd := &cobra.Command{
		Use:   "configure <provider>",
		Short: "Set provider configuration (api key, auth method, etc.)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			body := map[string]any{}
			if apiKey != "" {
				body["api_key"] = apiKey
			}
			if authMethod != "" {
				body["auth_method"] = authMethod
			}
			if extraJSON != "" {
				if err := mergeJSONInto(body, extraJSON); err != nil {
					return err
				}
			}
			if len(body) == 0 {
				return fmt.Errorf("at least one of --api-key, --auth-method, or --json is required")
			}
			return state.Client.ConfigureProvider(cmd.Context(), args[0], body)
		},
	}
	cmd.Flags().StringVar(&apiKey, "api-key", "", "API key value (kept secret server-side)")
	cmd.Flags().StringVar(&authMethod, "auth-method", "", "Active auth method id")
	cmd.Flags().StringVar(&extraJSON, "json", "", "Additional fields as a JSON object")
	return cmd
}

func newLLMDeleteConfigCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "delete-config <provider>",
		Short: "Remove all stored config for a provider",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			return state.Client.DeleteProviderConfig(cmd.Context(), args[0])
		},
	}
}

func newLLMModelGroupsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "model-groups",
		Short: "Read or write the high/low model group assignments",
	}
	cmd.AddCommand(
		&cobra.Command{
			Use:   "get",
			Short: "Show model group assignments",
			RunE: func(c *cobra.Command, _ []string) error {
				if err := requireToken(); err != nil {
					return err
				}
				out, err := state.Client.GetModelGroups(c.Context())
				if err != nil {
					return err
				}
				if state.Output.JSON {
					return output.PrintJSON(out)
				}
				dumpJSON(out)
				return nil
			},
		},
		newLLMModelGroupsSetCmd(),
	)
	return cmd
}

func newLLMModelGroupsSetCmd() *cobra.Command {
	var high, low, defaultProvider string
	cmd := &cobra.Command{
		Use:   "set",
		Short: "Update model group assignments",
		RunE: func(cmd *cobra.Command, _ []string) error {
			if err := requireToken(); err != nil {
				return err
			}
			body := map[string]any{}
			groups := map[string]any{}
			if high != "" {
				groups["high"] = high
			}
			if low != "" {
				groups["low"] = low
			}
			if len(groups) > 0 {
				body["model_groups"] = groups
			}
			if defaultProvider != "" {
				body["default_provider"] = defaultProvider
			}
			if len(body) == 0 {
				return fmt.Errorf("at least one of --high, --low, --default-provider is required")
			}
			return state.Client.UpdateModelGroups(cmd.Context(), body)
		},
	}
	cmd.Flags().StringVar(&high, "high", "", "Model id for the 'high' group")
	cmd.Flags().StringVar(&low, "low", "", "Model id for the 'low' group")
	cmd.Flags().StringVar(&defaultProvider, "default-provider", "", "Default provider name")
	return cmd
}

func newLLMDeviceCodeCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "device-code",
		Short: "GitHub Copilot device-code authentication flow",
	}
	cmd.AddCommand(
		&cobra.Command{
			Use:   "start",
			Short: "Start a device-code flow; prints user code + verification URL",
			RunE: func(c *cobra.Command, _ []string) error {
				resp, err := state.Client.DeviceCodeStart(c.Context())
				if err != nil {
					return err
				}
				if state.Output.JSON {
					return output.PrintJSON(resp)
				}
				output.PrintKV([][2]string{
					{"verification_uri", resp.VerificationURI},
					{"user_code", resp.UserCode},
					{"device_code", resp.DeviceCode},
					{"expires_in", fmt.Sprintf("%d", resp.ExpiresIn)},
					{"interval", fmt.Sprintf("%d", resp.Interval)},
				})
				return nil
			},
		},
		&cobra.Command{
			Use:   "poll <device_code>",
			Short: "Poll until the user completes the flow",
			Args:  cobra.ExactArgs(1),
			RunE: func(c *cobra.Command, args []string) error {
				ctx := c.Context()
				deviceCode := args[0]
				interval := 5 * time.Second
				for {
					select {
					case <-ctx.Done():
						return ctx.Err()
					default:
					}
					resp, err := state.Client.DeviceCodePoll(ctx, deviceCode)
					if err != nil {
						return err
					}
					switch resp.Status {
					case "complete":
						if state.Output.JSON {
							return output.PrintJSON(resp)
						}
						if resp.AccessToken != "" {
							output.PrintKV([][2]string{
								{"status", "complete"},
								{"access_token", resp.AccessToken},
							})
						} else {
							output.Println("complete (token stored server-side)")
						}
						return nil
					case "expired":
						return fmt.Errorf("device code expired; run `opa llm device-code start` again")
					case "error":
						return fmt.Errorf("device-code flow error: %s", resp.Error)
					case "pending":
						if resp.SlowDown {
							interval += 5 * time.Second
						}
						time.Sleep(interval)
					default:
						return fmt.Errorf("unexpected status %q", resp.Status)
					}
				}
			},
		},
	)
	return cmd
}

func numField(m map[string]any, k string) string {
	if v, ok := m[k]; ok {
		switch n := v.(type) {
		case float64:
			return fmt.Sprintf("%d", int(n))
		case int:
			return fmt.Sprintf("%d", n)
		default:
			return fmt.Sprintf("%v", v)
		}
	}
	return ""
}

func mergeJSONInto(target map[string]any, body string) error {
	var src map[string]any
	if err := jsonUnmarshalString(body, &src); err != nil {
		return fmt.Errorf("invalid --json: %w", err)
	}
	for k, v := range src {
		target[k] = v
	}
	return nil
}

File: app/cli/commands/llm.py
"""`openpa llm ...` — manage LLM providers, models, and model groups.

Mirrors `cli/cmd/llm.go`.
"""

from __future__ import annotations

import json as _json
import sys
import time
from typing import Any, Optional

import typer

from app.cli.commands._helpers import graceful_errors


llm_app = typer.Typer(
    name="llm",
    help="Manage LLM providers, models, and model groups.",
    no_args_is_help=True,
)
providers_app = typer.Typer(
    name="providers",
    help="List, configure, and inspect LLM providers.",
    no_args_is_help=True,
)
model_groups_app = typer.Typer(
    name="model-groups",
    help="Read or write the high/low model group assignments.",
    no_args_is_help=True,
)
device_code_app = typer.Typer(
    name="device-code",
    help="GitHub Copilot device-code authentication flow.",
    no_args_is_help=True,
)
llm_app.add_typer(providers_app, name="providers")
llm_app.add_typer(model_groups_app, name="model-groups")
llm_app.add_typer(device_code_app, name="device-code")


# ── providers ─────────────────────────────────────────────────────────────

@providers_app.command("list")
@graceful_errors
def providers_list(ctx: typer.Context) -> None:
    """List configured and available providers."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import list_llm_providers
    from app.cli.config import Config
    from app.cli.output import OutputMode, Table, print_json
    from app.cli.output.formatting import bool_field, string_field

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> list[dict[str, Any]]:
        async with Client(cfg) as client:
            return await list_llm_providers(client)

    providers = asyncio.run(_run())

    if mode.json:
        print_json(providers)
        return

    table = Table(mode, "NAME", "DISPLAY", "CONFIGURED", "MODELS", "ACTIVE_AUTH")
    for p in providers:
        table.add_row(
            string_field(p, "name"),
            string_field(p, "display_name"),
            bool_field(p, "configured", False),
            _num_field(p, "model_count"),
            string_field(p, "active_auth_method"),
        )
    table.render()


@providers_app.command("models")
@graceful_errors
def providers_models(
    ctx: typer.Context,
    provider: str = typer.Argument(..., help="Provider name."),
) -> None:
    """List models for a provider."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import get_provider_models
    from app.cli.config import Config
    from app.cli.output import OutputMode, Table, print_json
    from app.cli.output.formatting import string_field

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> dict[str, Any]:
        async with Client(cfg) as client:
            return await get_provider_models(client, provider)

    out = asyncio.run(_run())

    if mode.json:
        print_json(out)
        return

    models = out.get("models") if isinstance(out, dict) else None
    table = Table(mode, "ID", "NAME")
    if isinstance(models, list):
        for m in models:
            if isinstance(m, dict):
                table.add_row(string_field(m, "id"), string_field(m, "name"))
    table.render()


@providers_app.command("configure")
@graceful_errors
def providers_configure(
    ctx: typer.Context,
    provider: str = typer.Argument(..., help="Provider name."),
    api_key: Optional[str] = typer.Option(
        None, "--api-key", help="API key value (kept secret server-side)."
    ),
    auth_method: Optional[str] = typer.Option(
        None, "--auth-method", help="Active auth method id."
    ),
    extra_json: Optional[str] = typer.Option(
        None, "--json", help="Additional fields as a JSON object."
    ),
) -> None:
    """Set provider configuration (api key, auth method, etc.)."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import configure_provider
    from app.cli.config import Config

    body: dict[str, Any] = {}
    if api_key:
        body["api_key"] = api_key
    if auth_method:
        body["auth_method"] = auth_method
    if extra_json:
        try:
            extra = _json.loads(extra_json)
        except _json.JSONDecodeError as e:
            typer.echo(f"invalid --json: {e}", err=True)
            raise typer.Exit(code=1) from e
        if not isinstance(extra, dict):
            typer.echo("--json must be an object", err=True)
            raise typer.Exit(code=1)
        body.update(extra)

    if not body:
        typer.echo(
            "at least one of --api-key, --auth-method, or --json is required",
            err=True,
        )
        raise typer.Exit(code=1)

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await configure_provider(client, provider, body)

    asyncio.run(_run())


@providers_app.command("delete-config")
@graceful_errors
def providers_delete_config(
    ctx: typer.Context,
    provider: str = typer.Argument(..., help="Provider name."),
) -> None:
    """Remove all stored config for a provider."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import delete_provider_config
    from app.cli.config import Config

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await delete_provider_config(client, provider)

    asyncio.run(_run())


# ── model-groups ──────────────────────────────────────────────────────────

@model_groups_app.command("get")
@graceful_errors
def model_groups_get(ctx: typer.Context) -> None:
    """Show model group assignments."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import get_model_groups
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    cfg.require_token()

    async def _run() -> dict[str, Any]:
        async with Client(cfg) as client:
            return await get_model_groups(client)

    out = asyncio.run(_run())

    if mode.json:
        print_json(out)
    else:
        sys.stdout.write(_json.dumps(out, indent=2, ensure_ascii=False, default=str) + "\n")


@model_groups_app.command("set")
@graceful_errors
def model_groups_set(
    ctx: typer.Context,
    high: Optional[str] = typer.Option(None, "--high", help="Model id for the 'high' group."),
    low: Optional[str] = typer.Option(None, "--low", help="Model id for the 'low' group."),
    default_provider: Optional[str] = typer.Option(
        None, "--default-provider", help="Default provider name."
    ),
) -> None:
    """Update model group assignments."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import update_model_groups
    from app.cli.config import Config

    body: dict[str, Any] = {}
    groups: dict[str, Any] = {}
    if high:
        groups["high"] = high
    if low:
        groups["low"] = low
    if groups:
        body["model_groups"] = groups
    if default_provider:
        body["default_provider"] = default_provider

    if not body:
        typer.echo(
            "at least one of --high, --low, --default-provider is required",
            err=True,
        )
        raise typer.Exit(code=1)

    cfg: Config = ctx.obj["cfg"]
    cfg.require_token()

    async def _run() -> None:
        async with Client(cfg) as client:
            await update_model_groups(client, body)

    asyncio.run(_run())


# ── device-code ───────────────────────────────────────────────────────────

@device_code_app.command("start")
@graceful_errors
def device_code_start(ctx: typer.Context) -> None:
    """Start a device-code flow; prints user code + verification URL."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import device_code_start as _start
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json, print_kv

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]

    async def _run():
        async with Client(cfg) as client:
            return await _start(client)

    resp = asyncio.run(_run())

    if mode.json:
        print_json(resp.to_dict())
        return
    print_kv([
        ("verification_uri", resp.verification_uri),
        ("user_code", resp.user_code),
        ("device_code", resp.device_code),
        ("expires_in", str(resp.expires_in)),
        ("interval", str(resp.interval)),
    ])


@device_code_app.command("poll")
@graceful_errors
def device_code_poll(
    ctx: typer.Context,
    device_code: str = typer.Argument(..., help="Device code from `device-code start`."),
) -> None:
    """Poll until the user completes the flow."""
    import asyncio

    from app.cli.client._base import Client
    from app.cli.client.llm import device_code_poll as _poll
    from app.cli.config import Config
    from app.cli.output import OutputMode, print_json, print_kv

    cfg: Config = ctx.obj["cfg"]
    mode: OutputMode = ctx.obj["mode"]
    interval = 5

    async def _run():
        nonlocal interval
        async with Client(cfg) as client:
            while True:
                resp = await _poll(client, device_code)
                if resp.status == "complete":
                    return resp
                if resp.status == "expired":
                    raise RuntimeError(
                        "device code expired; run `openpa llm device-code start` again"
                    )
                if resp.status == "error":
                    raise RuntimeError(f"device-code flow error: {resp.error}")
                if resp.status == "pending":
                    if resp.slow_down:
                        interval += 5
                    await asyncio.sleep(interval)
                    continue
                raise RuntimeError(f"unexpected status '{resp.status}'")

    try:
        resp = asyncio.run(_run())
    except RuntimeError as e:
        typer.echo(str(e), err=True)
        raise typer.Exit(code=1) from e
    except KeyboardInterrupt:
        raise typer.Exit(code=130)

    if mode.json:
        print_json(resp.to_dict())
        return
    if resp.access_token:
        print_kv([("status", "complete"), ("access_token", resp.access_token)])
    else:
        sys.stdout.write("complete (token stored server-side)\n")


def _num_field(d: dict[str, Any], key: str) -> str:
    v = d.get(key)
    if isinstance(v, (int, float)) and not isinstance(v, bool):
        return str(int(v))
    if v is None:
        return ""
    return str(v)
